clients/src/main/java/org/apache/kafka/clients/consumer/internals/RequestFuture.java

```diff
@@ -37,86 +37,94 @@
+    private final AtomicReference<Object> result = new AtomicReference<>(INCOMPLETE_SENTINEL);
+    private final ConcurrentLinkedQueue<RequestFutureListener<T>> listeners = new ConcurrentLinkedQueue<>();
 
     /**
      * Check whether the response is ready to be handled
      * @return true if the response is ready, false otherwise
      */
     public boolean isDone() {
-        return isDone;
+        return result.get() != INCOMPLETE_SENTINEL;
     }
 
     /**
      * Get the value corresponding to this request (only available if the request succeeded)
-     * @return the value if it exists or null
+     * @return the value set in {@link #complete(Object)}
+     * @throws IllegalStateException if the future is not complete or failed
      */
+    @SuppressWarnings("unchecked")
     public T value() {
-        return value;
+        if (!succeeded())
+            throw new IllegalStateException("Attempt to retrieve value from future which hasn't successfully completed");
+        return (T) result.get();
     }
 
     /**
      * Check if the request succeeded;
      * @return true if the request completed and was successful
      */
     public boolean succeeded() {
-        return isDone && exception == null;
+        return isDone() && !failed();
     }
 
     /**
      * Check if the request failed.
      * @return true if the request completed with a failure
      */
     public boolean failed() {
-        return isDone && exception != null;
+        return result.get() instanceof RuntimeException;
     }
 
     /**
      * Check if the request is retriable (convenience method for checking if
      * the exception is an instance of {@link RetriableException}.
      * @return true if it is retriable, false otherwise
+     * @throws IllegalStateException if the future is not complete or completed successfully
      */
     public boolean isRetriable() {
-        return exception instanceof RetriableException;
+        return exception() instanceof RetriableException;
     }
 
     /**
      * Get the exception from a failed result (only available if the request failed)
-     * @return The exception if it exists or null
+     * @return the exception set in {@link #raise(RuntimeException)}
+     * @throws IllegalStateException if the future is not complete or completed successfully
      */
     public RuntimeException exception() {
-        return exception;
+        if (!failed())
+            throw new IllegalStateException("Attempt to retrieve exception from future which hasn't failed");
+        return (RuntimeException) result.get();
     }
 
     /**
      * Complete the request successfully. After this call, {@link #succeeded()} will return true
      * and the value can be obtained through {@link #value()}.
      * @param value corresponding value (or null if there is none)
+     * @throws IllegalStateException if the future has already been completed
      */
     public void complete(T value) {
```

Comment:
Sure. I thought about adding this, but figured that would be an unlikely usage. Still probably better to be safe.